File: client/src/game/entities/plane.ts
import * as PIXI from "pixi.js";
import { EntityType } from "../../network/game/EntityType";
import { DrawLayer, Entity, } from "../entity";
import { getTexture } from "../resources";

const DIRECTIONS = 256;

export enum PlaneType {
  ALBATROS,
  BRISTOL,
  FOKKER,
  JUNKERS,
  SALMSON,
  SOPWITH,
}

const planeTypeTextureMap = new Map<PlaneType, number>();
planeTypeTextureMap.set(PlaneType.ALBATROS, 4);
planeTypeTextureMap.set(PlaneType.JUNKERS, 5);
planeTypeTextureMap.set(PlaneType.FOKKER, 6);
planeTypeTextureMap.set(PlaneType.BRISTOL, 7);
planeTypeTextureMap.set(PlaneType.SALMSON, 8);
planeTypeTextureMap.set(PlaneType.SOPWITH, 9);


export class Plane extends Entity {
  clientX = 0;
  clientY = 0;
  type = EntityType.PLANE;
  direction = 0;
  planeType = PlaneType.ALBATROS;

  container = new PIXI.Container();
  sprite = new PIXI.Sprite(getTexture("plane6.gif"));

  constructor() {
    super();

    this.sprite.anchor.set(0.5); // center sprite origin

    this.container.addChild(this.sprite);

    this.sprite.zIndex = DrawLayer.Plane;
  }

  getContainer(): PIXI.Container {
    return this.container;
  }

  getRadians() {
    return Math.PI * 2 * this.direction / DIRECTIONS;
  }

  getTexture() {
    const str = "plane" + planeTypeTextureMap.get(this.planeType) + ".gif"
    return getTexture(str);
  }

  redraw() {
    this.sprite.texture = this.getTexture();
    //console.log(this.planeType, PlaneType[this.planeType]);

    const x = this.clientX;
    const y = this.clientY;
    this.sprite.position.set(x, y);

    const rad = this.getRadians();
    // console.log(rad, this.direction)
    this.sprite.rotation = this.getRadians();

  }

  destroy() { }
}

File: game/src/main/java/com/lentokonepeli/entities/Plane.java
package com.lentokonepeli.entities;

import com.lentokonepeli.Entity;
import com.lentokonepeli.EntityType;
import com.lentokonepeli.Tickable;
import com.lentokonepeli.network.NetType;
import com.lentokonepeli.network.NetProp;

public class Plane extends Entity implements Tickable {

    public enum PlaneType {
        ALBATROS,
        BRISTOL,
        FOKKER,
        JUNKERS,
        SALMSON,
        SOPWITH,
    }

    public enum PlaneState {
        FLYING,
        LANDING,
        LANDED,
        TAKING_OFF,
        FALLING,
        DODGING,
    }

    private static final int COORDINATE_RESOLUTION = 100;
    private static final int DIRECTIONS = 256;
    private static final int MAX_Y = -570;
    private static final int SKY_HEIGHT = 500;
    private static final int speedPerPixel = 100;

    private int x;
    private int y;
    private PlaneState state;

    private NetProp<Integer> clientX = new NetProp<>("x", NetType.i16);
    private NetProp<Integer> clientY = new NetProp<>("y", NetType.i16);
    private NetProp<Integer> direction = new NetProp<>("direction", NetType.u8);
    private NetProp<Integer> planeType = new NetProp<>("planeType", NetType.u8);

    public Plane() {
        super(EntityType.PLANE);

        props.add(clientX);
        props.add(clientY);
        props.add(direction);
        props.add(planeType);

        clientX.set(Integer.valueOf(250));
        clientY.set(Integer.valueOf(-350));
        direction.set(Integer.valueOf(256 / 4));
        planeType.set(PlaneType.FOKKER.ordinal());
    }

    private double getRadians(int direction) {
        return Math.PI * 2 * direction / DIRECTIONS;
    }

    public void tick(long deltaMS) {
        var dir = this.direction.get() + 1;
        if (dir >= DIRECTIONS) {
            this.direction.set(Integer.valueOf(0));
        } else {
            this.direction.set(Integer.valueOf(dir));
        }

        if (Math.random() > 0.98) {

            var idx = (int) Math.floor(Math.random() * 6);
            this.planeType.set(Integer.valueOf(idx));
        }
    }

}
